### ISSUE TITLE
IllegalStateException when calling stream().toList() on ClientSetProxy [API-1392]

### ISSUE THREAD

**Issue by bergander** [hazelcast/hazelcast#21017]
**Describe the bug**
If an ISet of type ClientSetProxy is modified while calling stream().toList() then an exception could occur. 

```
private void testStream() {
  final HazelcastInstance client = HazelcastClient.newHazelcastClient();
  final ISet<?> mySet = client.getSet("mySet");
  mySet.stream().toList();
}
```

In IteratorSpliterator.estimateSize() there's a call to collections.size() and since the collection is a ClientSetProxy a remote call is made to determine the size. But that size might be different comparing to the number of elements fetched in later calls if the set as been modified.

> java.lang.IllegalStateException(End size 99 is less than fixed size 100) 
> java.base/java.util.stream.Nodes$FixedNodeBuilder.end(Nodes.java:1240) 
> java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258) 
> java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485) 
> java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) 
> java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550) 
> java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260) 
> java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616) 
> java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622) 
> java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627) 
> com.company.testStream(Test.java:xxx)

One way of resolving this issue would be to implement _stream()_ in _ClientSetProxy_ so that the IteratorSpliterator will contain a reference to a local copy of the set:

```
@Override
Stream stream() {
  getAll().stream();
}
```

Same problem might exist in other classes, like _ClientListProxy_.

**Expected behavior**
No thrown exception

**To Reproduce**

**Additional context**
Tested with JDK 16 and Hazelcast v4.2.2 and v5.1.1.

**Comment by sancar** [comment]
Hi @bergander thanks for the report. We will try to fix this in next 5.2 release and backport it to 5.1.x and 4.2.x series in minumum. We accept contributions if you have time and interest as well.

**Comment by bergander** [comment]
@sancar I've created a PR. Overriding the spliterator() method should fix the problem since both Collection.stream() and Collection.parallelStrem() calls that method.
